From d10228d4de2c12e5b51257108ff07c0d040d2ce9 Mon Sep 17 00:00:00 2001
From: Lionel Xu <Lionel.Xu@freescale.com>
Date: Thu, 11 Aug 2011 15:48:08 +0800
Subject: [PATCH 0342/2463] ENGR00154727 ESAI: Resolve i2c access problem when
 there is no audio card

I2c device should not probe successfully when there is no such device
on the bus. This will make i2c access failure later.

Signed-off-by: Lionel Xu <R63889@freescale.com>
---
 sound/soc/codecs/cs42888.c |   31 +++++++++++++++++--------------
 1 files changed, 17 insertions(+), 14 deletions(-)

diff --git a/sound/soc/codecs/cs42888.c b/sound/soc/codecs/cs42888.c
index 2ccc9df..dcaa65e 100644
--- a/sound/soc/codecs/cs42888.c
+++ b/sound/soc/codecs/cs42888.c
@@ -809,20 +809,6 @@ static int cs42888_probe(struct snd_soc_codec *codec)
 	}
 	msleep(1);
 
-	/* Verify that we have a CS42888 */
-	val = snd_soc_read(codec, CS42888_CHIPID);
-	if (val < 0) {
-		pr_err("Device with ID register %x is not a CS42888", val);
-		return -ENODEV;
-	}
-	/* The top four bits of the chip ID should be 0000. */
-	if ((val & CS42888_CHIPID_ID_MASK) != 0x00) {
-		dev_err(codec->dev, "device is not a CS42888\n");
-		return -ENODEV;
-	}
-
-	dev_info(codec->dev, "hardware revision %X\n", val & 0xF);
-
 	/* The I2C interface is set up, so pre-fill our register cache */
 	ret = cs42888_fill_cache(codec);
 	if (ret < 0) {
@@ -908,6 +894,23 @@ static int cs42888_i2c_probe(struct i2c_client *i2c_client,
 {
 	struct cs42888_private *cs42888;
 	int ret;
+	int val;
+
+	/* Verify that we have a CS42888 */
+	val = i2c_smbus_read_byte_data(i2c_client, CS42888_CHIPID);
+	if (val < 0) {
+		pr_err("Device with ID register %x is not a CS42888", val);
+		return -ENODEV;
+	}
+	/* The top four bits of the chip ID should be 0000. */
+	if ((val & CS42888_CHIPID_ID_MASK) != 0x00) {
+		dev_err(&i2c_client->dev, "device is not a CS42888\n");
+		return -ENODEV;
+	}
+
+	dev_info(&i2c_client->dev, "found device at i2c address %X\n",
+		i2c_client->addr);
+	dev_info(&i2c_client->dev, "hardware revision %X\n", val & 0xF);
 
 	/* Allocate enough space for the snd_soc_codec structure
 	   and our private data together. */
-- 
1.7.7.4

